test/test_optim.py

```diff
@@ -115,12 +118,90 @@ def _test_derived_optimizers(self, device, dtype, optim_info, flag, reduced_prec
+            torch.rand(2, 3, dtype=torch.float16, device='cuda:1', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.bfloat16, device='cuda:1', requires_grad=True),
+            torch.randint(1024, (2, 3), dtype=torch.int64, device='cuda:1', requires_grad=False),
+        ]
+
+        for p in params:
+            if p.requires_grad:
+                p.grad = torch.rand_like(p, device=p.device, dtype=p.dtype)
+
+        kIterations = 7 if flag == "foreach" else 1
+        optim_inputs = optim_info.optim_inputs_func()
+        optim_cls = optim_info.optim_cls
+        for optim_input in optim_inputs:
+            updated_params, state = [], []
+            kwargs = deepcopy(optim_input.kwargs)
+            if kwargs.get("capturable", False) and str(device) == "cpu":
+                # capturable is not supported on CPU
+                continue
+            for flag_value in (False, True):
+                kwargs[flag] = flag_value
+                params_clone = []
+                for p in params:
+                    p_clone = p.clone().detach()
+                    if p.requires_grad:
+                        p_clone.requires_grad = True
+                        p_clone.grad = p.grad.clone().detach()
+                        params_clone.append(p_clone)
+
+                optimizer = optim_cls(params_clone, **kwargs)
+                for _ in range(kIterations):
+                    optimizer.step()
+
+                state.append(optimizer.state)
+                updated_params.append(params_clone)
+
+            og_state, new_state = state
+            for og_p, new_p in zip(updated_params[0], updated_params[1]):
+                # Increasing the tolerance as we are collating lots of ops together for optimizers and
+                # the designated tolerances are for single op only.
+                single_rtol, single_atol = torch.testing._comparison.get_tolerances(new_p.dtype, rtol=None, atol=None)
+                rtol = 5 * single_rtol
+                atol = 5 * single_atol
+
+                self.assertEqual(og_p, new_p, rtol=rtol, atol=atol)
+
+                # check that optimizer states are the same
+                og_p_state = og_state[og_p]
+                new_p_state = new_state[new_p]
+
+                for k in og_p_state:
+                    actual = new_p_state[k]
+                    self.assertEqual(og_p_state[k], actual, rtol=rtol, atol=atol)
+
+
     @skipMPS  # MPS doesn't support torch.float64, see https://github.com/pytorch/pytorch/issues/115350
     @optims([optim for optim in optim_db if "foreach" in optim.supported_impls], dtypes=[torch.float64])
     def test_foreach_matches_forloop(self, device, dtype, optim_info):
         self._test_derived_optimizers(device, dtype, optim_info, "foreach")
 
 
+    @onlyCUDA
+    @unittest.skipIf(not TEST_MULTIGPU, "only one GPU detected")
+    @optims([optim for optim in optim_db if "foreach" in optim.supported_impls])
```

Comment:
nit: If it is not an explicit goal to separate foreach and fused tests (which I guess is not since they call the exact same helper), then I think you can merge the fused and foreach into a single parametrized test and remove the indirection into _test_derived_optimizers_mixed_device_dtype ?

Reply:
done